### PR TITLE
enable backward pass computation and communication overlap by prefetching all gather

### PR DESCRIPTION
**Pull request by zhaojuanmao**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69282

It is observed from past traces that all gather ops are not triggered until current layer's backward pass starts to compute, also for some models previous layers' reduce scatter is scheduled before next layer's all gather ops, since all gather and reduce scatter are in the same nccl stream, this case could result in backward pass has no communication and computation overlap.

To explicitly make next layers' all gather scheduled while previous layers' backward computation is running, we can prefetch next layers' all gather full params.  This can help 1) both all gather and reduce scatter are overlapped with computation deterministically 2) only prefetch one layer's all gather full parameters, to avoid increasing too much memories.

The implementation borrowed the idea from https://github.com/facebookresearch/fairscale/pull/865, where forward graph order is recorded in the forward pass. 

In the backward pass, this PR prefetches all gather full parameter in current layer's pre-backward hook, instead of prefetching in current layer's post backward hook in https://github.com/facebookresearch/fairscale/pull/865. Also make sure all gather streams are synced properly.

Experiments showed 10% memory increase and 20% latency speed up for 1GB roberta model in a slow network environment.

Differential Revision: [D32787760](https://our.internmc.facebook.com/intern/diff/D32787760/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

### CODE REVIEW

File: torch/distributed/_fsdp/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -546,6 +549,8 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
         # All-gather full parameters, moving them to compute_device if
         # necessary.
         self._rebuild_full_params()
+        # Wait for all_gather full parameters to finish before computation
+        torch.cuda.current_stream().wait_stream(self._streams["all_gather"])
```
Comment (rohan-varma) [inline torch/distributed/_fsdp/fully_sharded_data_parallel.py:553]: Is this needed due to other changes in this PR, or is it a correctness fix?

File: torch/distributed/_fsdp/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -366,6 +366,8 @@ def _reset_lazy_init(self) -> None:
         """
         self._is_root: Optional[bool] = None
         self._streams: Dict[str, torch.cuda.Stream] = {}
+        self._fsdp_graph_order: List[nn.Module] = []
+        self._my_fsdp_idx_in_graph: Optional[int] = None
```
Comment (rohan-varma) [inline torch/distributed/_fsdp/fully_sharded_data_parallel.py:370]: Can we have a single structure such as Dict[idx, nn.Module]?

File: torch/distributed/_fsdp/fully_sharded_data_parallel.py [hunk 3]
```diff
@@ -604,7 +613,20 @@ def _pre_backward_hook(*unused: Any) -> None:
...
             self._rebuild_full_params()
+            # Wait for all_gather to finish before computation
+            torch.cuda.current_stream().wait_stream(self._streams["all_gather"])
+
+            # Prefetch previous layer's full params in backward pass
+            if (
+                self._fsdp_graph_order is not None
+                and self._my_fsdp_idx_in_graph is not None and self._my_fsdp_idx_in_graph > 0
+            ):
+                self._fsdp_graph_order[self._my_fsdp_idx_in_graph - 1]._rebuild_full_params()  # type: ignore[operator]
```
Comment (rohan-varma) [inline torch/distributed/_fsdp/fully_sharded_data_parallel.py:629]: what happens to the existing `self._rebuild_full_params()` calls? Is there a risk we call them redundantly for the same module, or is this handled somehow?

File: torch/distributed/_fsdp/fully_sharded_data_parallel.py [hunk 4]
```diff
@@ -900,8 +922,6 @@ def update_p_data(output_tensor: torch.Tensor) -> None:
                     # Set p.data = output_tensor (with padding trimmed)
                     update_p_data(output_tensor)
 
-        torch.cuda.current_stream().wait_stream(self._streams["all_gather"])
```
Comment (rohan-varma) [inline torch/distributed/_fsdp/fully_sharded_data_parallel.py:903]: Why is this removed?

**Review by rohan-varma — COMMENTED** [review]
Thanks for this performance improvement! I had a couple of questions about it.

File: torch/distributed/_fsdp/fully_sharded_data_parallel.py (hunk 2 above)
Comment (zhaojuanmao) [reply torch/distributed/_fsdp/fully_sharded_data_parallel.py:370]: we need List structure to preserve the order of layers

File: torch/distributed/_fsdp/fully_sharded_data_parallel.py (hunk 1 above)
Comment (zhaojuanmao) [reply torch/distributed/_fsdp/fully_sharded_data_parallel.py:553]: yes, it is needed, because I removed all_gather stream sync from _rebuild_full_params(), so we need to add it here.  all_gather stream sync is removed from _rebuild_full_params() because all gather does not need to be waited by current computation stream during prefetching call

File: torch/distributed/_fsdp/fully_sharded_data_parallel.py (hunk 3 above)
Comment (zhaojuanmao) [reply torch/distributed/_fsdp/fully_sharded_data_parallel.py:629]: it is handled inside _rebuild_full_params()

File: torch/distributed/_fsdp/fully_sharded_data_parallel.py (hunk 4 above)
Comment (zhaojuanmao) [reply torch/distributed/_fsdp/fully_sharded_data_parallel.py:903]: see above comments

File: torch/distributed/_fsdp/fully_sharded_data_parallel.py [hunk 5]
```diff
@@ -554,6 +559,10 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
 
         outputs = self.module(*args, **kwargs)
 
+        if self not in self._fsdp_graph_order:
```
Comment (rohan-varma) [inline torch/distributed/_fsdp/fully_sharded_data_parallel.py:562]: This order is populated just once, is it possible that graph order changes because of logic in forward pass by the user?

**Review by rohan-varma — COMMENTED** [review]
Awesome, thanks! 

One question in the comments, and also, is it worth it to have a flag to enable/disable this because of 10% memory increase? 10% memory increase could potentially limit the size of the model that can be trained right?

**Review by rohan-varma — APPROVED** [review]

**Comment by zhaojuanmao** [comment]
Have fixed a few corner cases for prefetching full parameters in post backward hook.  

After benchmarking, prefetching full parameters in the pre-backward hook has the best performance and stable but at cost of increased memory; prefetching full parameters in the post-backward hook did not see expected performance, also failed in a few corner cases (fixed) although there is no memory increase. The main issue is that post backward hook fire order is not consistent with opposite of forward computation order, so incorrectly prefetched all gather could delay the really needed all gather in the single NCCL stream and cause some layer's computation delay. 

So will put these two algorithms as two configurable experimental algorithms for now

**Comment by zhaojuanmao** [comment]
creating a new PR, closing this
